### PR TITLE
AaveV3LeverageStrategyExtension audit feedback

### PR DESCRIPTION
**Pull request by ckoopmann**
Verifying and fixing issues identified by auditors in the AaveV3LeverageStrategyExtension contract.

Issues to be fixed:
- [x] [Getting wrong ltv for assets in emode](https://github.com/sherlock-audit/2023-05-Index-judging/issues/251) -[fix](https://github.com/IndexCoop/index-coop-smart-contracts/pull/142/commits/5526d84f2da21999854c515be0ce5b013cfe1058)
- [x] [Adjust calculation of maxCollateral](https://github.com/sherlock-audit/2023-05-Index-judging/issues/254#issuecomment-1601070157) - [fix](https://github.com/IndexCoop/index-coop-smart-contracts/pull/142/commits/718dd42baf2095a911c6325dcd5ca6553d4d8977)
- [x] [Add validation of targetLeverageRatio > 1](https://github.com/sherlock-audit/2023-05-Index-judging/issues/313) - [fix](https://github.com/IndexCoop/index-coop-smart-contracts/pull/142/commits/eeb34fbd3beb83338e878a0b844ac14ee7fd2ad7)
- [x] [Change order of mul / div to avoid precision loss](https://github.com/sherlock-audit/2023-05-Index-judging/issues/299) - [fix1](https://github.com/IndexCoop/index-coop-smart-contracts/pull/142/commits/9d04e7bfcd82b157577d79fd2fcb227accef82de) [fix2](https://github.com/IndexCoop/index-coop-smart-contracts/pull/142/commits/0e291b490f21e986a8a384c398bdf189ca96410e)
- [x] [Replace use of deprecated `latestAnswer` (optional)](https://github.com/sherlock-audit/2023-05-Index-judging/issues/187) - [fix](https://github.com/IndexCoop/index-coop-smart-contracts/pull/142/commits/7ab6d28a5e2028b6fcca1f8571242f5f30c873f5)
- [x] [Use AaveOracle instead of individually configured Chainlink oracles](https://github.com/sherlock-audit/2023-05-Index-judging/issues/323) - [fix](https://github.com/IndexCoop/index-coop-smart-contracts/pull/142/commits/bed0e348e05e17e3c0e11be4cf6c22bf900abd36)

Issues not to be fixed
- [Add keeper specified slippage parameter](https://github.com/sherlock-audit/2023-05-Index-judging/issues/285)

### CODE REVIEW

File: contracts/adapters/AaveV3LeverageStrategyExtension.sol [hunk 1]
```diff
@@ -19,9 +19,15 @@
...
-import {AaveLeverageStrategyExtension} from "./AaveLeverageStrategyExtension.sol";
+import { Math } from "@openzeppelin/contracts/math/Math.sol";
 
+import {AaveLeverageStrategyExtension} from "./AaveLeverageStrategyExtension.sol";
 import {IBaseManager} from "../interfaces/IBaseManager.sol";
+import { IChainlinkAggregatorV3 } from "../interfaces/IChainlinkAggregatorV3.sol";
+import {IPool} from "../interfaces/IPool.sol";
+import {IPoolAddressesProvider} from "../interfaces/IPoolAddressesProvider.sol";
+import { DataTypes } from "../interfaces/Datatypes.sol";
+import { PreciseUnitMath } from "../lib/PreciseUnitMath.sol";
```
Comment (FlattestWhite) [inline contracts/adapters/AaveV3LeverageStrategyExtension.sol:30]: Formatting nits

**Review by FlattestWhite — APPROVED** [review]

File: contracts/adapters/AaveV3LeverageStrategyExtension.sol [hunk 2]
```diff
@@ -31,14 +37,22 @@ import {IBaseManager} from "../interfaces/IBaseManager.sol";
...
         IBaseManager _manager,
         ContractSettings memory _strategy,
         MethodologySettings memory _methodology,
         ExecutionSettings memory _execution,
         IncentiveSettings memory _incentive,
         string[] memory _exchangeNames,
-        ExchangeSettings[] memory _exchangeSettings
+        ExchangeSettings[] memory _exchangeSettings,
+        IPoolAddressesProvider _lendingPoolAddressesProvider,
+        uint256 _maxOraclePriceAge
```
Comment (ckoopmann) [inline contracts/adapters/AaveV3LeverageStrategyExtension.sol:55]: In the deployment we will have to set this parameter to a reasonable value. Meaning the maximum "age" / "delay" of price updates to be expected on chainlink.
@FlattestWhite @pblivin0x

File: contracts/adapters/AaveV3LeverageStrategyExtension.sol [hunk 3]
```diff
@@ -66,4 +106,116 @@ contract AaveV3LeverageStrategyExtension is AaveLeverageStrategyExtension {
...
+            return _actionInfo.collateralBalance
+                .preciseMul(netRepayLimit.sub(_actionInfo.borrowValue))
+                .preciseDiv(netRepayLimit);
+        }
+    }
+
+    function _getLtvAndLiquidationThreshold() internal view returns(uint256, uint256) {
+        if(currentEModeCategoryId != 0 ) {
+            // Retrieve collateral factor and liquidation threshold for the collateral asset in precise units (1e16 = 1%)
+            DataTypes.EModeCategory memory emodeData = IPool(lendingPoolAddressesProvider.getPool()).getEModeCategoryData(currentEModeCategoryId);
```
Comment (pblivin0x) [inline contracts/adapters/AaveV3LeverageStrategyExtension.sol:137]: 👀 update for e-mode data https://github.com/sherlock-audit/2023-05-Index-judging/issues/251

File: contracts/adapters/AaveV3LeverageStrategyExtension.sol [hunk 4]
```diff
@@ -66,4 +106,116 @@ contract AaveV3LeverageStrategyExtension is AaveLeverageStrategyExtension {
...
+        if (_isLever) {
+            uint256 netBorrowLimit = _actionInfo.collateralValue
+                .preciseMul(maxLtvRaw.mul(10 ** 14))
+                .preciseMul(PreciseUnitMath.preciseUnit().sub(execution.unutilizedLeveragePercentage));
+
+            return netBorrowLimit
+                .sub(_actionInfo.borrowValue)
+                .preciseDiv(_actionInfo.collateralPrice);
+        } else {
+            uint256 netRepayLimit = _actionInfo.collateralValue
```
Comment (pblivin0x) [inline contracts/adapters/AaveV3LeverageStrategyExtension.sol:125]: 👀 update to loosen restriction of `netRepayLimit` https://github.com/sherlock-audit/2023-05-Index-judging/issues/254

File: contracts/adapters/AaveV3LeverageStrategyExtension.sol [hunk 5]
```diff
@@ -66,4 +106,116 @@ contract AaveV3LeverageStrategyExtension is AaveLeverageStrategyExtension {
...
+        MethodologySettings memory _methodology,
+        ExecutionSettings memory _execution,
+        IncentiveSettings memory _incentive
+    )
+        internal
+        override
+        pure
+    {
+        super._validateNonExchangeSettings(_methodology, _execution, _incentive);
+        require(_methodology.targetLeverageRatio >= 1 ether, "Target leverage ratio must be >= 1e18");
```
Comment (pblivin0x) [inline contracts/adapters/AaveV3LeverageStrategyExtension.sol:156]: ✅ added check https://github.com/sherlock-audit/2023-05-Index-judging/issues/313

File: contracts/adapters/AaveV3LeverageStrategyExtension.sol [hunk 6]
```diff
@@ -66,4 +106,116 @@ contract AaveV3LeverageStrategyExtension is AaveLeverageStrategyExtension {
...
+    }
+
+    function _calculateMinRepayUnits(uint256 _collateralRebalanceUnits, uint256 _slippageTolerance, ActionInfo memory _actionInfo) internal override pure returns (uint256) {
+        return _collateralRebalanceUnits
+            .preciseMul(_actionInfo.collateralPrice)
+            .preciseMul(PreciseUnitMath.preciseUnit().sub(_slippageTolerance)) // Changed order of mul / div here
+            .preciseDiv(_actionInfo.borrowPrice);
```
Comment (pblivin0x) [inline contracts/adapters/AaveV3LeverageStrategyExtension.sol:163]: ✅ updated order https://github.com/sherlock-audit/2023-05-Index-judging/issues/299

File: contracts/adapters/AaveV3LeverageStrategyExtension.sol [hunk 7]
```diff
@@ -66,4 +106,116 @@ contract AaveV3LeverageStrategyExtension is AaveLeverageStrategyExtension {
...
+        rebalanceInfo.borrowBalance = strategy.targetBorrowDebtToken.balanceOf(address(strategy.setToken));
+        rebalanceInfo.collateralValue = rebalanceInfo.collateralPrice.preciseMul(rebalanceInfo.collateralBalance);
+        rebalanceInfo.borrowValue = rebalanceInfo.borrowPrice.preciseMul(rebalanceInfo.borrowBalance);
+        rebalanceInfo.setTotalSupply = strategy.setToken.totalSupply();
+
+        return rebalanceInfo;
+    }
+
+    function _getAssetPrice(IChainlinkAggregatorV3 _priceOracle, uint256 _decimalAdjustment) internal view returns (uint256) {
+        (,int256 rawPrice,, uint256 updatedAt,) = _priceOracle.latestRoundData();
```
Comment (pblivin0x) [inline contracts/adapters/AaveV3LeverageStrategyExtension.sol:214]: ✅ `latestRoundData` https://github.com/sherlock-audit/2023-05-Index-judging/issues/187

File: contracts/adapters/AaveV3LeverageStrategyExtension.sol [hunk 8]
```diff
@@ -42,9 +42,11 @@ contract AaveV3LeverageStrategyExtension is AaveLeverageStrategyExtension {
     IPoolAddressesProvider public lendingPoolAddressesProvider;
     uint256 public maxOraclePriceAge;
     bool public overrideNoRebalanceInProgress;
+    IAaveOracle public aaveOracle;
 
     constructor(
         IBaseManager _manager,
+        IAaveOracle _aaveOracle,
```
Comment (FlattestWhite) [inline contracts/adapters/AaveV3LeverageStrategyExtension.sol:49]: The aaveOracle should come from calling `getPriceOracle` on the `LendingPoolAddressProvider`

**Review by pblivin0x — CHANGES_REQUESTED** [review]
All changes here LGTM 💯

Before final approval and merge, lets make sure all dependencies in our integration tests are up to date with the index-protocol audit fixes https://github.com/IndexCoop/index-protocol/commit/520a9cf7bcfd9b898d990f8d20589ea2a97a2886
- [ ] `AaveV3LeverageModule`
- [ ] `DebtIssuanceModuleV2`
- [ ] `Invoke`

File: contracts/adapters/AaveV3LeverageStrategyExtension.sol (hunk 8 above)
Comment (ckoopmann) [reply contracts/adapters/AaveV3LeverageStrategyExtension.sol:49]: Good point, thanks. Will adjust 👍

Comment (ckoopmann) [reply contracts/adapters/AaveV3LeverageStrategyExtension.sol]: Adjusted in:
https://github.com/IndexCoop/index-coop-smart-contracts/pull/142/commits/5d1cc5945ae49511d3ee546a0cccadad47fbd111

Comment (ckoopmann) [reply contracts/adapters/AaveV3LeverageStrategyExtension.sol]: In fact I went one step further and just read the address anew from the addressProvider everytime when I call the oracle.
Thereby we don't get rugged if for some reason the AaveOracle address changes:

https://github.com/IndexCoop/index-coop-smart-contracts/pull/142/commits/0a69def8d226b8921e98ad88c3158cb70ead0df6

File: contracts/adapters/AaveV3LeverageStrategyExtension.sol (hunk 1 above)
Comment (ckoopmann) [reply contracts/adapters/AaveV3LeverageStrategyExtension.sol:30]: Fixed with a bunch of other formatting / documentation related changes in this Commit:
https://github.com/IndexCoop/index-coop-smart-contracts/pull/142/commits/3b5c204925aca0c0c31ee847d2788372b404ab7a

**Comment by ckoopmann** [comment]
> All changes here LGTM 💯
> 
> Before final approval and merge, lets make sure all dependencies in our integration tests are up to date with the index-protocol audit fixes [IndexCoop/index-protocol@520a9cf](https://github.com/IndexCoop/index-protocol/commit/520a9cf7bcfd9b898d990f8d20589ea2a97a2886)
> 
> * [ ]  `AaveV3LeverageModule`
> * [ ]  `DebtIssuanceModuleV2`
> * [ ]  `Invoke`

In the integration tests for the StrategyExtension I am testing against deployed contracts. (DebtIssuanceModule, SetTokenCreator etc.)
Therefore to make sure those are testing against the correct versions with the fixed invoke issue, we need to redeploy those first.

**Comment by pblivin0x** [comment]
> > All changes here LGTM 💯
> > Before final approval and merge, lets make sure all dependencies in our integration tests are up to date with the index-protocol audit fixes [IndexCoop/index-protocol@520a9cf](https://github.com/IndexCoop/index-protocol/commit/520a9cf7bcfd9b898d990f8d20589ea2a97a2886)
> > 
> > * [ ]   `AaveV3LeverageModule`
> > * [ ]   `DebtIssuanceModuleV2`
> > * [ ]   `Invoke`
> 
> In the integration tests for the StrategyExtension I am testing against deployed contracts. (DebtIssuanceModule, SetTokenCreator etc.) Therefore to make sure those are testing against the correct versions with the fixed invoke issue, we need to redeploy those first.
> 
> Basically we will have to update the addresses [here](https://github.com/IndexCoop/index-coop-smart-contracts/blob/3b5c204925aca0c0c31ee847d2788372b404ab7a/test/integration/ethereum/aaveV3LeverageStrategyExtension.spec.ts#L75) and also adjust the test such that it just uses the new A3LM deployment instead of redeploying the module.

Changes implemented here #145

**Comment by ckoopmann** [comment]
> > > All changes here LGTM 💯
> > > Before final approval and merge, lets make sure all dependencies in our integration tests are up to date with the index-protocol audit fixes [IndexCoop/index-protocol@520a9cf](https://github.com/IndexCoop/index-protocol/commit/520a9cf7bcfd9b898d990f8d20589ea2a97a2886)
> > > 
> > > * [ ]    `AaveV3LeverageModule`
> > > * [ ]    `DebtIssuanceModuleV2`
> > > * [ ]    `Invoke`
> > 
> > 
> > In the integration tests for the StrategyExtension I am testing against deployed contracts. (DebtIssuanceModule, SetTokenCreator etc.) Therefore to make sure those are testing against the correct versions with the fixed invoke issue, we need to redeploy those first.
> > Basically we will have to update the addresses [here](https://github.com/IndexCoop/index-coop-smart-contracts/blob/3b5c204925aca0c0c31ee847d2788372b404ab7a/test/integration/ethereum/aaveV3LeverageStrategyExtension.spec.ts#L75) and also adjust the test such that it just uses the new A3LM deployment instead of redeploying the module.
> 
> Changes implemented here #145

Noted that in the linked PR only the A3LM json-file / bytecode is updated. I guess technically we would also want to update the DIM and other modules that use the invoke library. But this is nothing that should necessarily block us from merging this I guess.

**Review by snake-poison — APPROVED** [review]
LGTM!

**Pull request merged**